src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -95,6 +104,50 @@ public static Task<List<string>> ReadRequestAndSendResponseAsync(Socket server,
             return lines;
         }
 
+        public static async Task<bool> WebSocketHandshakeAsync(Socket s, StreamReader reader, StreamWriter writer)
+        {
+            string serverResponse = null;
+            string currentRequestLine;
+            while (!string.IsNullOrEmpty(currentRequestLine = await reader.ReadLineAsync().ConfigureAwait(false)))
+            {
+                string[] tokens = currentRequestLine.Split(new char[] { ':' }, 2);
+                if (tokens.Length == 2)
+                {
+                    string headerName = tokens[0];
+                    if (headerName == "Sec-WebSocket-Key")
+                    {
+                        string headerValue = tokens[1].Trim();
+                        string responseSecurityAcceptValue = ComputeWebSocketHandshakeSecurityAcceptValue(headerValue);
+                        serverResponse =
+                            "HTTP/1.1 101 Switching Protocols\r\n" +
+                            "Upgrade: websocket\r\n" +
+                            "Connection: Upgrade\r\n" +
+                            "Sec-WebSocket-Accept: " + responseSecurityAcceptValue + "\r\n\r\n";
+                    }
+                }
+            }
+
+            if (serverResponse != null)
+            {
+                // We received a valid WebSocket opening handshake. Send the appropriate response.
+                await writer.WriteAsync(serverResponse).ConfigureAwait(false);
+                return true;
+            }
+
+            return false;
+        }
+
+        private static string ComputeWebSocketHandshakeSecurityAcceptValue(string secWebSocketKey)
+        {
+            // GUID specified by RFC 6455.
+            const string Rfc6455Guid = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
+            string combinedKey = secWebSocketKey + Rfc6455Guid;
+
+            SHA1 sha1Provider = new SHA1CryptoServiceProvider();
```

Comment:
You should add a comment above this line saying something like "Use of SHA1 hash is required by RFC 6455." Otherwise, this might get flag in static analysis tools as a security issue.  SHA1 is no longer favored.  For example, SHA256 is used for crypto certs etc.